pkg/filesystem/local_directory_windows.go

```diff
@@ -439,50 +439,60 @@ func (d *localDirectory) Mknod(name path.Component, perm os.FileMode, deviceNumb
 	return status.Error(codes.Unimplemented, "Creation of device nodes is not supported on Windows")
 }
 
+func clamp(value, max uint32) uint32 {
+	if value > max {
+		return max
+	}
+
+	return value
+}
+
 func readdirnames(handle windows.Handle) ([]string, error) {
-	outBufferSize := uint32(512)
-	outBuffer := make([]byte, outBufferSize)
-	firstIteration := true
+	outBufferSize := uint32(256)
+	names := make([]string, 0)
+
 	for {
-		err := windows.GetFileInformationByHandleEx(handle, windows.FileFullDirectoryInfo,
-			&outBuffer[0], outBufferSize)
-		if err == nil {
-			break
-		}
-		if err.(syscall.Errno) == windows.ERROR_NO_MORE_FILES {
-			if firstIteration {
-				return []string{}, nil
+		// NB: In testing this was found to be about the largest buffer the API
+		// would fill. Any larger allocation would be unused and lead to
+		// resource exhaustion. We have not found good documentation for how to
+		// use this.
+		outBufferSize = clamp(outBufferSize*2, 62000)
```

Comment:
Instead of doing this, why don't we simply use a buffer that is twice as big as what the last call to `GetFileInformationByHandleEx()` **filled**? So if we give it a 4 KB buffer and the kernel fills the entire 4 KB, then the next time we call it with 8 KB. Then we don't need to make guesses about what the kernel's limit actually is.